Mathlib/Combinatorics/Enumerative/DyckWord.lean

```diff
@@ -0,0 +1,187 @@
+  l := p.l.take i
+  bal := hi
+  nonneg k := by rw [take_take]; exact p.nonneg (min k i)
+
+/-- Suffix of a Dyck word as a Dyck word, given that the count of `U`s and `D`s in the prefix
+are equal. -/
+def drop (i : ℕ) (hi : (p.l.take i).count U = (p.l.take i).count D) : DyckWord where
+  l := p.l.drop i
+  bal := by
+    have := p.bal
+    rw [← take_append_drop i p.l, count_append, count_append] at this
+    omega
+  nonneg k := by
+    rw [show i = min i (i + k) by omega, ← take_take] at hi
+    rw [take_drop, ← add_le_add_iff_left (((p.l.take (i + k)).take i).count U),
+      ← count_append, hi, ← count_append, take_append_drop]
+    exact p.nonneg _
+
+/-- Nest `p` in one pair of brackets, i.e. `x` becomes `(x)`. -/
+def nest : DyckWord where
+  l := [U] ++ p.l ++ [D]
+  bal := by simp [p.bal]
+  nonneg i := by
+    simp only [take_append_eq_append_take, count_append]
+    rw [← add_rotate (count D _), ← add_rotate (count U _)]
+    apply add_le_add _ (p.nonneg _)
+    cases' i.eq_zero_or_pos with hi hi; · simp [hi]
+    rw [take_all_of_le (show [U].length ≤ i by rwa [length_singleton]), count_singleton']
+    simp only [ite_true, ite_false]
+    rw [add_comm]
+    exact add_le_add (zero_le _) ((count_le_length _ _).trans (by simp))
+
+/-- Denest `p`, i.e. `(x)` becomes `x`, given that `p` is strictly positive in its interior
+(this ensures that `x` is a Dyck word). -/
+def denest (h : p.l ≠ []) (pos : ∀ i, 0 < i → i < p.l.length →
+    (p.l.take i).count D < (p.l.take i).count U) : DyckWord where
+  l := p.l.dropLast.tail
+  bal := by
+    have := p.bal
+    rw [← p.cons_tail_dropLast_concat h, count_append, count_cons] at this
+    simpa using this
+  nonneg i := by
+    have l1 : p.l.take 1 = [p.l.head h] := by rcases p with - | - <;> tauto
+    have l3 : p.l.length.pred = p.l.length.pred.pred + 1 := by
+      rcases p with - | ⟨s, ⟨- | ⟨t, r⟩⟩⟩
+      · tauto
+      · rename_i bal _
+        cases s <;> simp at bal
+      · tauto
+    rw [← drop_one, take_drop, dropLast_eq_take, take_take]
+    have ub : min (1 + i) p.l.length.pred < p.l.length :=
+      (min_le_right _ p.l.length.pred).trans_lt (Nat.pred_lt ((length_pos.mpr h).ne'))
+    have lb : 0 < min (1 + i) p.l.length.pred := by
+      rw [l3, add_comm, min_add_add_right]; omega
+    have eq := pos _ lb ub
+    set j := min (1 + i) p.l.length.pred
+    rw [← (p.l.take j).take_append_drop 1, count_append, count_append, take_take,
+      min_eq_left (by omega), l1, p.head_eq_U] at eq
+    simp only [count_singleton', ite_true, ite_false] at eq
+    omega
+
+lemma denest_nest (h : p.l ≠ []) (pos : ∀ i, 0 < i → i < p.l.length →
+    (p.l.take i).count D < (p.l.take i).count U) : (p.denest h pos).nest = p := by
+  simpa [DyckWord.ext_iff] using p.cons_tail_dropLast_concat h
+
+section Semilength
+
+/-- The semilength of a Dyck word is half of the number of `DyckStep`s in it, or equivalently
+its number of `U`s. -/
+def semilength : ℕ := p.l.count U
```

Comment:
Can we have a lemma showing it’s equal to the count of D?

Reply:
All done.